src/controller/tests/data_model/DataModelFixtures.h

```diff
@@ -0,0 +1,98 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *  @file
+ *  This module provides shared fixture implementations of global functions
+ *  for data model tests as well as global variables to control them.
+ */
+
+#pragma once
+
+#include <app/CommandHandler.h>
+#include <app/util/mock/Constants.h>
+#include <app/util/mock/Functions.h>
+#include <lib/core/DataModelTypes.h>
+
+namespace chip {
+namespace app {
+namespace DataModelTests {
+
+constexpr EndpointId kTestEndpointId = 1;
+
+constexpr ClusterId kPerpetualClusterId     = Test::MockClusterId(2);
+constexpr AttributeId kPerpetualAttributeid = Test::MockAttributeId(1);
+
+constexpr DataVersion kRejectedDataVersion = 1;
+constexpr DataVersion kAcceptedDataVersion = 5;
+constexpr DataVersion kDataVersion         = kAcceptedDataVersion;
+
+constexpr uint8_t kExampleClusterSpecificSuccess = 11u;
+constexpr uint8_t kExampleClusterSpecificFailure = 12u;
+
+constexpr ClusterStatus kTestSuccessClusterStatus = 1;
+constexpr ClusterStatus kTestFailureClusterStatus = 2;
+
+enum class ReadResponseDirective
+{
+    kSendDataResponse,
+    kSendManyDataResponses,          // Many data blocks, for a single concrete path
+                                     // read, simulating a malicious server.
+    kSendManyDataResponsesWrongPath, // Many data blocks, all using the wrong
+                                     // path, for a single concrete path
+                                     // read, simulating a malicious server.
+    kSendDataError,
+    kSendTwoDataErrors, // Multiple errors, for a single concrete path,
+                        // simulating a malicious server.
+};
+extern ReadResponseDirective gReadResponseDirective;
+
+// Number of reads of Clusters::UnitTesting::Attributes::Int16u that we have observed.
+// Every read will increment this count by 1 and return the new value.
+extern uint16_t gInt16uTotalReadCount;
+
+extern bool gIsLitIcd;
+
+enum class WriteResponseDirective
+{
+    kSendAttributeSuccess,
+    kSendAttributeError,
+    kSendMultipleSuccess,
+    kSendMultipleErrors,
+    kSendClusterSpecificSuccess,
+    kSendClusterSpecificFailure,
+};
+extern WriteResponseDirective gWriteResponseDirective;
```

Comment:
these seem to control how tests behave - changing this may have a sideffect for other tests.

Should we maybe use https://github.com/project-chip/connectedhomeip/blob/master/src/lib/support/Scoped.h to wrap these and force scoped changes only within tests?

Reply:
These globals existed before but were only changeable from on of the test files each (though potentially affecting the other tests). I’ll check if all modifications fit with the scoped pattern.

Reply:
I've made these use `ScopedChangeOnly` now. I added the ability to assign to a `ScopedChange` because some tests made further changes within the tests not directly scoped to a block.

I think this is in keeping with the `ScopedChange` pattern since the key thing is that the value is restored at the end of the scope; what happens within the scope is not important.